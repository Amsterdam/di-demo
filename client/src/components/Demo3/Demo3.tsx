import React, { useState, useEffect } from 'react';
import createIrmaSession from '@services/createIrmaSession';
import content from '@services/content';
import ReactMarkDown from 'react-markdown';
import * as AscLocal from '@components/LocalAsc/LocalAsc';
import { Accordion } from '@amsterdam/asc-ui';
import { Alert as AlertIcon } from '@amsterdam/asc-assets';
import CredentialSelector, { CredentialSource } from '@components/CredentialSelector/CredentialSelector';
import ExternalLink from '@components/ExternalLink/ExternalLink';
import PageTemplate from '@components/PageTemplate/PageTemplate';
import BreadCrumbs from '@components/BreadCrumbs';
import DemoNotification from '@components/DemoNotification/DemoNotification';
import ResponsiveImage, { IHeaderImageProps } from '@components/ResponsiveImage/ResponsiveImage';
import QRCode from '@components/QRCode/QRCode';
import EmphasisBlock from '@components/EmphasisBlock/EmphasisBlock';
import { Checkmark } from '@amsterdam/asc-assets';
import ContentBlock from '@components/ContentBlock/ContentBlock';
import WhyIRMA from '@components/WhyIRMA/WhyIRMA';
import preloadDemoImages from '@services/preloadImages';
import { startSurvey as startUsabillaSurvey } from '@services/usabilla';

export interface IProps {}
// @todo add error flow with incorrect data

const Demo3: React.FC<IProps> = () => {
    const [credentialSource, setCredentialSource] = useState(CredentialSource.PRODUCTION);
    const [hasResult, setHasResult] = useState<boolean>(false);
    const [hasError, setHasError] = useState<boolean>(false);
    // const [bsn, setBsn] = useState<string>('');
    const [name, setName] = useState<string>('');

    const getSession = async () => {
        const response = await createIrmaSession(
            'demo3',
            'irma-qr',
            credentialSource === CredentialSource.DEMO && { demo: true }
        );
        if (response) {
            setHasResult(true);
            setHasError(false);
            // setBsn(response['bsn']);
            setName(response['fullname']);
        } else {
            setHasError(true);
        }
        window.scrollTo(0, 0);
        startUsabillaSurvey();
        return response;
    };

    // Define dynamic header image
    const [headerImg, setHeaderImg] = useState<IHeaderImageProps>({
        filename: content.responsiveImages.demo3.header.src,
        alt: content.responsiveImages.demo3.header.alt
    });

    // Preload demo images
    useEffect(() => {
        preloadDemoImages(
            Object.keys(content.responsiveImages.demo3).map(key => content.responsiveImages.demo3[key].src)
        );
    }, []);

    // Update header image for 18+
    useEffect(() => {
        if (hasResult) {
            setHeaderImg({
                filename: content.responsiveImages.demo3.headerResult.src,
                alt: content.responsiveImages.demo3.headerResult.alt
            });
        }
    }, [hasResult]);

    return (
        <PageTemplate>
            <ContentBlock>
                <CredentialSelector credentialSource={credentialSource} setCredentialSource={setCredentialSource} />
                {!hasResult && !hasError && <DemoNotification />}
                <ReactMarkDown
                    source={content.demo3.breadcrumbs}
                    renderers={{ list: BreadCrumbs, listItem: BreadCrumbs.Item }}
                />

                <ReactMarkDown
                    source={content.demo3[hasResult ? 'proven' : 'unproven'].title}
                    renderers={{ heading: AscLocal.H1 }}
                />

                {hasResult && !hasError && (
                    <AscLocal.Alert
                        color={AscLocal.AlertColor.SUCCESS}
                        icon={<Checkmark />}
                        iconSize={14}
                        heading={content.demo3.proven.alert.title}
                        content={content.demo3.proven.alert.body.replace(/\[\]/, name)}
                        dataTestId="hasResultAlert"
                    />
                )}

                {hasError && (
                    <AscLocal.Alert
                        color={AscLocal.AlertColor.ERROR}
                        icon={<AlertIcon />}
                        iconSize={22}
                        heading={content.demoErrorAlert.heading}
                        content={content.demoErrorAlert.content}
                        dataTestId="hasErrorAlert"
                    />
                )}
            </ContentBlock>

            <ResponsiveImage filename={headerImg.filename} alt={headerImg.alt} />

            {!hasResult ? (
                <AscLocal.Row noMargin>
                    <AscLocal.Column
                        span={{
                            small: 1,
                            medium: 2,
                            big: 6,
                            large: 9,
                            xLarge: 9
                        }}
                    >
                        <ContentBlock>
                            <section>
                                <ReactMarkDown
                                    source={content.demo3.unproven.intro}
                                    renderers={{
                                        heading: AscLocal.H2,
                                        paragraph: AscLocal.Paragraph,
                                        list: AscLocal.UL
                                    }}
                                />
                            </section>
                            <section>
                                <AscLocal.AccordionContainer>
                                    <Accordion title={content.demo3.unproven.why.title}>
                                        <ReactMarkDown
                                            source={content.demo3.unproven.why.body}
                                            renderers={{
                                                paragraph: AscLocal.Paragraph,
                                                list: AscLocal.UL
                                            }}
                                        />
                                    </Accordion>
                                </AscLocal.AccordionContainer>
                            </section>
                            <section>
                                <QRCode getSession={getSession} label={content.demo3.button} />
                            </section>
                            <section>
                                <ReactMarkDown
                                    source={content.downloadIrma}
                                    renderers={{
                                        paragraph: AscLocal.Paragraph,
                                        link: ExternalLink
                                    }}
                                />
                            </section>
                        </ContentBlock>
                    </AscLocal.Column>
                    <AscLocal.Column
                        span={{
                            small: 1,
                            medium: 2,
                            big: 6,
                            large: 3,
                            xLarge: 3
                        }}
                    >
                        <WhyIRMA />
                    </AscLocal.Column>
                </AscLocal.Row>
            ) : (
                <>
                    <ContentBlock>
                        <section>
                            <ReactMarkDown source={content.noSavePromise} />
                        </section>
                    </ContentBlock>
                    <EmphasisBlock>
                        <ContentBlock>
                            <section>
                                <ReactMarkDown
                                    source={content.demo3.result}
                                    renderers={{
                                        heading: AscLocal.H2,
                                        paragraph: AscLocal.Paragraph,
                                        list: AscLocal.UL
                                    }}
                                />
                            </section>
                            <section>
                                <ReactMarkDown
                                    source={content.callToAction}
                                    renderers={{
                                        heading: AscLocal.H2,
                                        paragraph: AscLocal.Paragraph,
                                        list: AscLocal.UL,
                                        link: AscLocal.InlineLink
                                    }}
                                />
                            </section>
                        </ContentBlock>
                    </EmphasisBlock>
                </>
            )}
        </PageTemplate>
    );
};

export default Demo3;
