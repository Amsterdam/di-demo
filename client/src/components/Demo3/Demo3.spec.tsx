import React from 'react';
import axios from 'axios';
import Demo3 from './Demo3';
import { withAppContext } from './../../test/utils';
import { render, fireEvent, act, queryByAltText } from '@testing-library/react';
import MockAdapter from 'axios-mock-adapter';

const mock = new MockAdapter(axios);

describe.only('Demo3 component', () => {
    it('should render correctly in demo demo and result page', async () => {
        mock.onGet('/config').reply(200, {
            irma: 'irma-server'
        });

        mock.onGet('/getsession/demo3?demo=true').reply(200, {
            sessionPtr: {
                u: 'https://acc.attr.auth.amsterdam.nl/irma/session/IikJk1eDBZV9iMb79gR3',
                irmaqr: 'disclosing'
            },
            token: '6aH2xaaLtUKgx3PknxDA'
        });

        // {  bsn: '123456789',
        // fullname: 'Willen van Zanten'
        // });

        // 'https://acc.attr.auth.amsterdam.nl/irma/session/pFqbViPJrkohP2KUR1fZ/status'

        // 'https://acc.attr.auth.amsterdam.nl/session/mfeXW6zqF1gdtpF1hMeX/result'
        // {"token":"mfeXW6zqF1gdtpF1hMeX","status":"DONE","type":"disclosing","proofStatus":"VALID","disclosed":[[{"rawvalue":"J.G. Swart","value":{"":"J.G. Swart","en":"J.G. Swart","nl":"J.G. Swart"},"id":"pbdf.gemeente.personalData.fullname","status":"PRESENT","issuancetime":1591228800},{"rawvalue":"191822668","value":{"":"191822668","en":"191822668","nl":"191822668"},"id":"pbdf.gemeente.personalData.bsn","status":"PRESENT","issuancetime":1591228800}]]}

        // 'https://acc.attr.auth.amsterdam.nl/session/mfeXW6zqF1gdtpF1hMeX/result'
        // {"token":"mfeXW6zqF1gdtpF1hMeX","status":"DONE","type":"disclosing","proofStatus":"VALID","disclosed":[[{"rawvalue":"J.G. Swart","value":{"":"J.G. Swart","en":"J.G. Swart","nl":"J.G. Swart"},"id":"pbdf.gemeente.personalData.fullname","status":"PRESENT","issuancetime":1591228800},{"rawvalue":"191822668","value":{"":"191822668","en":"191822668","nl":"191822668"},"id":"pbdf.gemeente.personalData.bsn","status":"PRESENT","issuancetime":1591228800}]]}

        const { container, asFragment, debug, queryAllByText, getByTestId } = render(withAppContext(<Demo3 />));

        expect(queryAllByText('Demo 3: Inloggen met IRMA').length).toEqual(2);

        // expect(asFragment()).toMatchSnapshot();

        fireEvent.click(getByTestId('qrCodeButton'));

        act(() => {
            expect(queryAllByText('Demo 3: Inloggen met IRMA').length).toEqual(2);
        });
    });
});
