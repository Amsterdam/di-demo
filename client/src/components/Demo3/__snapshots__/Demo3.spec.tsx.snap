// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Demo3 should render the initial header image 1`] = `
.c0 {
  width: 100%;
  margin-bottom: 8px;
}

@media screen and (min-width:1200px) {

}

@media screen and (min-width:1024px) {

}

@media screen and (min-width:1024px) {

}

@media screen and (min-width:540px) {

}

@media screen and (min-width:1024px) {

}

@media (min-width:1200px) {

}

@media (min-width:1024px) and (max-width:1199px) {

}

@media (min-width:700px) and (max-width:1023px) {

}

@media (min-width:414px) and (max-width:699px) {

}

@media (max-width:413px) {

}

@media (min-width:1200px) {

}

@media (min-width:1024px) and (max-width:1199px) {

}

@media (min-width:700px) and (max-width:1023px) {

}

@media (min-width:414px) and (max-width:699px) {

}

@media (max-width:413px) {

}

@media (min-width:1200px) {

}

@media (min-width:1024px) and (max-width:1199px) {

}

@media (min-width:700px) and (max-width:1023px) {

}

@media (min-width:414px) and (max-width:699px) {

}

@media (max-width:413px) {

}

@media (min-width:1200px) {

}

@media (min-width:1024px) and (max-width:1199px) {

}

@media (min-width:700px) and (max-width:1023px) {

}

@media (min-width:414px) and (max-width:699px) {

}

@media (max-width:413px) {

}

@media (min-width:1200px) {

}

@media (min-width:1024px) and (max-width:1199px) {

}

@media (min-width:700px) and (max-width:1023px) {

}

@media (min-width:414px) and (max-width:699px) {

}

@media (max-width:413px) {

}

@media (min-width:1200px) {

}

@media (min-width:1024px) and (max-width:1199px) {

}

@media (min-width:700px) and (max-width:1023px) {

}

@media (min-width:414px) and (max-width:699px) {

}

@media (max-width:413px) {

}

@media screen and (min-width:540px) {

}

@media screen and (max-width:1199px) {

}

@media screen and (min-width:375px) {

}

@media screen and (max-width:1199px) {

}

@media screen and (min-width:1200px) {

}

@media (max-width:1199px) {

}

@media (min-width:768px) {

}

@media (min-width:1024px) {

}

@media (min-width:1200px) {

}

@media screen and (min-width:768px) {

}

@media (max-width:1199px) {

}

@media (min-width:1200px) {

}

@media (min-width:1200px) {

}

@media (max-width:767px) {

}

@media (min-width:1024px) {

}

@media (min-width:1024px) {

}

@media (min-width:1024px) {

}

<img
  alt=""
  class="c0"
  data-testid="headerImage"
  src="/assets/probeer-irma-940.jpg"
  srcset="/assets/probeer-irma-290.jpg 580w, /assets/probeer-irma-940.jpg 1880w"
/>
`;

exports[`Demo3 should update the page after completing the IRMA flow 1`] = `
.c0 {
  width: 100%;
  margin-bottom: 8px;
}

@media screen and (min-width:1200px) {

}

@media screen and (min-width:1024px) {

}

@media screen and (min-width:1024px) {

}

@media screen and (min-width:540px) {

}

@media screen and (min-width:1024px) {

}

@media (min-width:1200px) {

}

@media (min-width:1024px) and (max-width:1199px) {

}

@media (min-width:700px) and (max-width:1023px) {

}

@media (min-width:414px) and (max-width:699px) {

}

@media (max-width:413px) {

}

@media (min-width:1200px) {

}

@media (min-width:1024px) and (max-width:1199px) {

}

@media (min-width:700px) and (max-width:1023px) {

}

@media (min-width:414px) and (max-width:699px) {

}

@media (max-width:413px) {

}

@media (min-width:1200px) {

}

@media (min-width:1024px) and (max-width:1199px) {

}

@media (min-width:700px) and (max-width:1023px) {

}

@media (min-width:414px) and (max-width:699px) {

}

@media (max-width:413px) {

}

@media (min-width:1200px) {

}

@media (min-width:1024px) and (max-width:1199px) {

}

@media (min-width:700px) and (max-width:1023px) {

}

@media (min-width:414px) and (max-width:699px) {

}

@media (max-width:413px) {

}

@media (min-width:1200px) {

}

@media (min-width:1024px) and (max-width:1199px) {

}

@media (min-width:700px) and (max-width:1023px) {

}

@media (min-width:414px) and (max-width:699px) {

}

@media (max-width:413px) {

}

@media (min-width:1200px) {

}

@media (min-width:1024px) and (max-width:1199px) {

}

@media (min-width:700px) and (max-width:1023px) {

}

@media (min-width:414px) and (max-width:699px) {

}

@media (max-width:413px) {

}

@media screen and (min-width:540px) {

}

@media screen and (max-width:1199px) {

}

@media screen and (min-width:375px) {

}

@media screen and (max-width:1199px) {

}

@media screen and (min-width:1200px) {

}

@media (max-width:1199px) {

}

@media (min-width:768px) {

}

@media (min-width:1024px) {

}

@media (min-width:1200px) {

}

@media screen and (min-width:768px) {

}

@media (max-width:1199px) {

}

@media (min-width:1200px) {

}

@media (min-width:1200px) {

}

@media (max-width:767px) {

}

@media (min-width:1024px) {

}

@media (min-width:1024px) {

}

@media (min-width:1024px) {

}

<img
  alt=""
  class="c0"
  data-testid="headerImage"
  src="/assets/mijn-amsterdam-940.jpg"
  srcset="/assets/mijn-amsterdam-290.jpg 580w, /assets/mijn-amsterdam-940.jpg 1880w"
/>
`;
